cmd/flux/create_secret_notation.go

```
@@ -0,0 +1,187 @@
+	Example: ` # Create a Notation configuration secret on disk and encrypt it with Mozilla SOPS
+  flux create secret notation config \
+    --namespace=my-namespace \
+    --trust-policy-file=./my-trust-policy.json \
+    --ca-cert-file=my-./my-cert.crt \
+	--ca-cert-file=my-./my-cert.pem \
+    --export > repo-auth.yaml
+
+  sops --encrypt --encrypted-regex '^(data|stringData)$' \
+    --in-place repo-auth.yaml`,
+
+	RunE: createSecretNotationCmdRun,
+}
+
+type secretNotationFlags struct {
+	trustPolicyFile string
+	caCrtFile       []string
+}
+
+var secretNotationArgs secretNotationFlags
+
+func init() {
+	createSecretNotationCmd.Flags().StringVar(&secretNotationArgs.trustPolicyFile, "trust-policy-file", "", "notation trust policy file path")
+	createSecretNotationCmd.Flags().StringSliceVar(&secretNotationArgs.caCrtFile, "ca-cert-file", []string{}, "root ca cert file or directory path")
+
+	createSecretCmd.AddCommand(createSecretNotationCmd)
+}
+
+func createSecretNotationCmdRun(cmd *cobra.Command, args []string) error {
+	if len(args) < 1 {
+		return fmt.Errorf("name is required")
+	}
+
+	if secretNotationArgs.caCrtFile == nil || len(secretNotationArgs.caCrtFile) == 0 {
+		return fmt.Errorf("--ca-cert-file cannot be empty")
+	}
+
+	if secretNotationArgs.trustPolicyFile == "" {
+		return fmt.Errorf("--trust-policy-file cannot be empty")
+	}
+
+	name := args[0]
+
+	labels, err := parseLabels()
+	if err != nil {
+		return err
+	}
+
+	policy, err := os.ReadFile(secretNotationArgs.trustPolicyFile)
+	if err != nil {
+		return fmt.Errorf("unable to read trust policy file: %w", err)
+	}
+
+	var doc trustpolicy.Document
+
+	if err := json.Unmarshal(policy, &doc); err != nil {
+		return fmt.Errorf("error occurred while parsing %s: %w", secretNotationArgs.trustPolicyFile, err)
+	}
+
+	if err := doc.Validate(); err != nil {
+		return fmt.Errorf("trust policy validation failed: %w", err)
```

```suggestion
		return fmt.Errorf("invalid trust policy: %w", err)
```

---

On this I'm in two minds about it. In [source-controller](https://github.com/fluxcd/source-controller/blob/014365d0e13ecee62174f67bcaf4724fa65f8cc2/internal/oci/notation/notation.go#L201-L217) we check for verification level `skip` and ensure `TrustStores` and `TrustedIdentities` are empty.  In situations where `TrustStores` and `TrustedIdentities` are populated and `VerificationLevel` is set to `skip`, `doc.Validate()` will return an error.

We do check the `trust-policy.json` is a valid json which might be enough but at the same time it might be nice to also ensure the trust policy we're supplying is actually valid. Perhaps adding a flag to fail if the validation returns an error otherwise log it?

What do you think, @souleb?

---

I think we should return the error as is. Diverging from `notation` in any way could be a pain in the future. The user should have the same experience that it would get with `notation`.

Also This is just a helper. The users that want to switch back to `skip` with existing `TrustStores` and `TrustedIdentities` can do it manually.

---

Works for me! Resolved in [9160c2a](https://github.com/fluxcd/flux2/pull/4735/commits/9160c2aac0be213fe3d1e774301c58343d14b8a5).